Use live apphost when publishing ilc as singlefile

am11 opened the pull request:
Using the following command to cross-build product for FreeBSD:
```sh
docker run --rm -v$(pwd)/runtime:/runtime -e ROOTFS_DIR=/crossrootfs/x64 \
  mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-cross-amd64-freebsd-14 \
  sh -c '/runtime/build.sh clr+libs+packs -c Debug -os freebsd -cross &&
  tdnf install -y file unzip && cd /tmp &&
  unzip /runtime/artifacts/packages/Debug/Shipping/runtime.freebsd-x64.Microsoft.DotNet.ILCompiler.9.0.0-dev.nupkg &&
  find . -name ilc -exec file {} \;'
```

before:

> ./tools/ilc: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=771fdbe539f221e5b1e4e6dc000846f10d503b3f, stripped

after:

> ./tools/ilc: ELF 64-bit LSB pie executable, x86-64, version 1 (FreeBSD), dynamically linked, interpreter /libexec/ld-elf.so.1, for FreeBSD 14.0 (1400097), FreeBSD-style, BuildID[sha1]=4b265d82eb3e1de5315251f7a5db2713ced94d2e, stripped

---

It required defer building ILCompiler.csproj until after apphost sfxproj. Also, PackageRID is used to restore the "host" package to build the target, while OutputRID is used to create package for the target, which is what we need here.

Fix #104497

am11 committed:
Use live apphost when publishing ilc as singlefile

jkotas commented on src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj:
```diff
@@ -1,18 +1,28 @@
-    <RuntimeIdentifier>$(PackageRID)</RuntimeIdentifier>
+    <RuntimeIdentifier>$(OutputRID)</RuntimeIdentifier>
+
+    <!-- Mobile platforms don't have apphost and we don't need it. However, restore requires
```
We are not shipping naot that runs on phones. For example, https://nuget.info/packages/runtime.ios-arm64.Microsoft.DotNet.ILCompiler/8.0.0-preview.6.23329.7 contains target libraries, but no compiler. The compiler is expected to come from the host package (e.g. osx-arm64).

Would it be better to stop building ILCompiler for mobile instead of these workarounds?

am11 replied:
Good point. I will give it a try. :)

am11 committed:
Block ILCompiler.csproj build on mobile

am11 committed:
Invert condition

jkotas commented on eng/Subsets.props:
```diff
@@ -529,13 +528,17 @@
 
   <!-- Packs sets -->
 
+  <ItemGroup>
```
The tricky ordering of the different build steps looks fragile.

I am wondering whether it would be better to use multiple build.sh invocations to bootstrap the FreeBSD cross build. The first invocation would skip building parts that depend on last-known-good packages for target arch, the second invocation would then build the rest and use the packages produced by the first invocation.

am11 replied:
Yup, it is fragile as in, when we are not building packs subset, it fails the build on platforms using PublishSingleFile. Pulling Microsoft.NETCore.App.Host.sfxproj in clr subset is also not an option because that depends on libs+host to build first.

There is an existing bug that on the systems where we use PublishAot for ILCompiler, it still looks for apphost (https://github.com/dotnet/runtime/pull/80154#issuecomment-2060649529, it was actually a regression in .net9 p1). That makes the refactoring those parts a degree more delicate.

am11 replied:
I've changed it so that ordering only affect cross-OS builds. In that, it doesn't build the incorrect ilc on FreeBSD if they are not building packs subset.

am11 replied:
I've changed it so that ordering only affect cross-OS builds. In that, it doesn't build the incorrect ilc on FreeBSD if packs subset is not built.

am11 replied:
@jkotas, the ILCompiler.csproj build will be now deferred until after packs are built, only for freebsd/illumos/haiku OS set which do not provide Microsoft SDK. This requires single invocation of `./build.sh`.

The current state of this PR ([351856c](https://github.com/dotnet/runtime/pull/105004/commits/351856c73283c6b451aec83c055e4decf8030931)) is such that if we still don't like Subsets.props changes, then we can revert Subsets.props (alone) and document the following steps for the devs of those platforms:

```sh
# invocation 1 - build everything without ILCompiler support
$ ./build.sh -os freebsd -p:NativeAotSupported=false

# invocation 2 - build clr.tools and packs subsets again without NativeAotSupported
#                condition so ILCompiler.csproj can use live-built apphost.
$ ./build.sh clr.tools+packs -os freebsd
```

jkotas replied:
Ideally, the cross-arch and cross-OS bootstrapping builds would produce a full-featured bits that are equivalent (including how the tools get published) to the natively compiled packages. I think achieving it via multiple invocations of build.sh flow would be easier to maintain than the ordering of the individual build steps.

am11 replied:
@Thefrank, @sec, basically this is the new workflow for community platforms for cross-OS build (e.g. for freebsd on linux): https://github.com/dotnet/runtime/pull/105004#discussion_r1737107996 (two steps). I tested it with this workflow https://github.com/am11/CrossRepoCITesting/actions/runs/10845732989/workflow, and artifacts are https://github.com/am11/CrossRepoCITesting/releases/tag/freebsd_10845732989. The `ilc` executable in ` runtime.freebsd-x64.Microsoft.DotNet.ILCompiler.10.0.0-ci.nupkg` shows:

```sh
$ unzip ~/Downloads/runtime.freebsd-x64.Microsoft.DotNet.ILCompiler.10.0.0-ci.nupkg
$ find . -name ilc -exec file {} \; | fold
./tools/ilc: ELF 64-bit LSB pie executable, x86-64, version 1 (FreeBSD), dynamic
ally linked, interpreter /libexec/ld-elf.so.1, for FreeBSD 14.0 (1400097), FreeB
SD-style, BuildID[sha1]=235f8e6220cbf08ec8c1966a415f82f6ed072e4e, with debug_inf
o, not stripped
```

jkotas replied:
> $ ./build.sh -os freebsd -p:NativeAotSupported=false
> $ ./build.sh clr.tools+packs -os freebsd

I think we need to have a clean separation of responsibilities between the two commands. Otherwise, I expect it will collide with where https://github.com/dotnet/runtime/pull/107772 is trying to go. A good split may be:

- The first command should be partial build that produces packages that the full build expects to come from LKG toolsets when there is one for target platform
- The second command should be a full build that uses the packages produced by the first command

We can define a new subset for the first command to make the command line UX simpler and stable over time. This subset would not be on by default. Building this subset can drop the packages in a new directory. The full build can automatically pick the LKG toolset from there if the directory exists.

Ideally, it would be even possible for the first command to be debug/checked build and the second command to be release build or vice versa, and it should work for non-cross builds too. 

@agocke Thoughts?

am11 committed:
Limit the scope of build order

am11 committed:
Merge branch 'main' into feature/freebsd-port/outputrid

agocke commented:
FYI, we're having some internal discussions about the live build as a whole. This is a large amount of complexity, vs always using the LKG build, and we're not sure it's worth it. There's a decent chance we back out the live build entirely and just use LKG bits for AOT'ing tools in the product.

am11 commented:
@agocke thanks. This particular change is for those platforms where:
* it is using cross-OS build (freebsd on linux)
* `UseNativeAotForComponents` is false
* apphost is needed to publish ILCompiler, and apphost being used is for linux, instead of the just-built freebsd.

I think this is a bit special mode, under which only community-supported platforms are currently falling (freebsd, illumos, haiku).

am11 committed:
Merge branch 'main' into feature/freebsd-port/outputrid

am11 committed:
Address CR feedback

am11 committed:
Merge dotnet/main into feature/freebsd-port/outputrid

am11 commented on src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj:
```diff
@@ -88,15 +91,6 @@
-      <SingleFileHostSourcePath></SingleFileHostSourcePath>
-    </PropertyGroup>
-  </Target>
-  <!-- END: Workaround for https://github.com/dotnet/runtime/issues/67742 -->
```
Yup, I've deleted it.

am11 commented on src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj:
```diff
@@ -88,15 +91,6 @@
   <Target Name="PublishCompiler"
```
Yup, it's redundant.

am11 commented on src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj:
```diff
@@ -1,18 +1,20 @@
     <OutputPath>$(RuntimeBinDir)ilc/</OutputPath>
-    <RuntimeIdentifier>$(PackageRID)</RuntimeIdentifier>
+    <RuntimeIdentifier>$(OutputRID)</RuntimeIdentifier>
+    <UseLocalTargetingRuntimePack Condition="'$(UseNativeAotForComponents)' != 'true' and '$(CrossBuild)' == 'true' and '$(TargetOS)' != '$(HostOS)'">true</UseLocalTargetingRuntimePack>
```
```suggestion
    <UseLocalTargetingRuntimePack Condition="'$(UseNativeAotForComponents)' != 'true' and '$(CrossBuild)' == 'true' and '$(TargetOS)' != '$(HostOS)' and '$(TargetsMobile)' != 'true'">true</UseLocalTargetingRuntimePack>
```

am11 committed:
Update src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj

am11 committed:
Update RID

am11 committed:
Use first step in CI (for now)

am11 commented on eng/pipelines/runtime.yml:
```diff
@@ -519,7 +519,7 @@ extends:
             testScope: innerloop
             nameSuffix: CoreCLR
-            buildArgs: -s clr+libs+host+packs -c $(_BuildConfig) -rc Checked
+            buildArgs: -s clr+libs+host+packs -c $(_BuildConfig) -rc Checked -p:NativeAotSupported=false
```
@jkoritzinsky, do we have any precedence where we have multi-steps script? We want to run two steps for freebsd: https://github.com/dotnet/runtime/pull/105004#discussion_r1737107996 (step 2 can be enabled later, but just wanted to highlight that we should have `./build.sh clr.tools+packs -os freebsd` coverage at some point as well)

jkoritzinsky replied:
Generally we use `postBuildSteps` to put the pieces together here, possibly using the `global-build-step` template if we need to run the root build script multiple separate times.

jkoritzinsky replied:
If you look at the runtime-official pipeline, we have a number of examples.

am11 replied:
Thank you! Now both steps are in place: https://dev.azure.com/dnceng-public/public/_build/results?buildId=808130&view=logs&j=c2b2f27a-5591-5c35-747e-70b670f3ca20&t=79e2e633-dd91-54d5-0384-9375c8c35f7d :)

am11 committed:
Add postBuildSteps for step 2

am11 committed:
Merge branch 'main' into feature/freebsd-port/outputrid

am11 committed:
fix

jkotas commented on src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj:
```diff
@@ -1,18 +1,21 @@
   <PropertyGroup>
     <OutputPath>$(RuntimeBinDir)ilc/</OutputPath>
-    <RuntimeIdentifier>$(PackageRID)</RuntimeIdentifier>
+    <RuntimeIdentifier Condition="'$(_IsPublishing)' != 'true'">$(PackageRID)</RuntimeIdentifier>
```
Where is the `_IsPublishing` property set? 

@agocke's comment https://github.com/dotnet/runtime/pull/107772#issuecomment-2350008766 makes it sound like the `_IsPublishing` is not set as expected in this project.

agocke replied:
Right now the dotnet CLI sets `_IsPublishing` when running `dotnet publish` but it's everyone else's responsibility to set it (I don't like this behavior, fyi). As far as I can tell, ILCompiler doesn't guarantee this is set, currently.

agocke commented:
> I think we need to have a clean separation of responsibilities between the two commands. Otherwise, I expect it will collide with where https://github.com/dotnet/runtime/pull/107772 is trying to go. A good split may be:

Sorry, I'm not really following the difference between those two commands.

My expectation, looking at it from a naive perspective, is that the difference between the first and second commands is primarily that `-p:NativeAotSupported=false` is set -- which to me would indicate that everything that might use Native AOT does the same thing, but instead uses trimmed single-file + R2R.

I don't really understand the "second build uses the first build" part.

jkotas commented:
> looking at it from a naive perspective, is that the difference between the first and second commands is primarily that -p:NativeAotSupported=false is set -- which to me would indicate that everything that might use Native AOT does the same thing, but instead uses trimmed single-file + R2R.

`NativeAotSupported` property controls whether to build native AOT packages. It does not directly control whether to AOT compile tools during the build that is controlled by `UseNativeAotForComponents` property. `UseNativeAotForComponents` is [false](https://github.com/dotnet/runtime/blob/d58b1cade30a6aaf94d187fa0d5f939c3e561e29/eng/Subsets.props#L125
) when cross-building from Linux to FreeBSD, so nothing will get AOT compiled when executing these commands. Also, these commands are made to work by strategically placed conditions in the build scripts, like [this one](https://github.com/dotnet/runtime/blob/d58b1cade30a6aaf94d187fa0d5f939c3e561e29/src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj#L24-L25).

Let me state the problem that I am trying to solve: We agreed to use LKG runtime and AOT compilers for all components that #107772 is about. Also, we said that it would be nice to have an option to use live runtime and AOT compilers as an option that this PR is about. What should be the UX for that option?

My proposal is that it should be a sequence of two build.cmd/.sh script invocations. The first invocation is a special build flow that builds just the runtime and AOT compiler bits that come from LKG by default. The second invocation is a regular build flow except that it uses runtime and AOT compilers produced by the first invocation instead of the LKG one.

am11 commented:
@jkotas @agocke, this issue is solving https://github.com/dotnet/runtime/issues/104497 for freebsd (with no pre-built SDK in Microsoft feed). The issue is about wrong "apphost package" (linux) is being used for freebsd (and not wrong ILCompiler package). First solution was to change the order of subset and use live apphost without breaking invocation,  it worked (without breaking existing platforms) but changing the subset order was not clean. Now we have split the two steps; first build the product without publishing ILCompiler. In second round, once the apphost package is built in first invocation, we use that one to publish ILCompiler project.

Please CMIIW, but I think LKG change is not improving #104497 scenario, right?

crossgen2_publish.csproj is a good separation which splits publishing and building concerns so package restore (PackageRID) and package creation (OutputRIID) are kept separate. If we switch ILCompiler to the same scheme, then we won't be needing some of these twists and turns.

jkotas commented:
I agree with you that live build is better in some situations, like enabling new platforms. Live build is worse for day-to-day development. 

crossgen2 publishing added to test build script in #106965 takes a while on debug builds. It made day-to-day development experience worse that triggered a discussion on our Teams channel about why we are using live build for crossgen2. This PR is outcome of that discussion. 

We use a mix of LKG builds and live builds for tools today. The goal of this PR is to consistently use LKG build everywhere by default.

We should have an option to use live build. Enabling this option should switch all tools to use live build and it should not change other settings to make things work - for example, it should not turn off PublishAot. My comment at https://github.com/dotnet/runtime/pull/105004#issuecomment-2350770931 is a proposal for how this can work.

HTH

am11 commented:
@jkotas, thanks for the explanation. The first invocation with `NativeAotSupported=<default for platform>` will always produce the incorrect binary with `--os freebsd --cross` (linux ELF instead of freebsd ELF https://github.com/dotnet/runtime/issues/104497). The second invocation will first need to undo the first invocation's work like evicting the cache (stuff in obj/ and passing /t:Rebuild) so the project doesn't get skipped by msbuild.

Wouldn't it be better to skip publishing the incorrect binary in the first place (the current state of this PR)?

jkotas commented:
I think that the option to use live build for tools should be an explicit gesture. Here are examples of scenarios that I would like to see it handle (open to suggestion for a better name of `-buildtoolsruntime`):

- Bootstrap community supported platform:
```
./build.sh -os freebsd -c release -buildtoolsruntime
./build.sh -os freebsd -c release
```

- Produce release build of the AOT compilers that run on live checked build of the runtime (useful for debugging purposes):
```
./build.sh -c checked -buildtoolsruntime
./build.sh -c release
```

> The first invocation with NativeAotSupported=<default for platform> will always produce the incorrect binary

`-buildtoolsruntime` in my example does not actually need to produce the native FreeBSD tools. The native FreeBSD tools are not used by the second step. 

One problem is that the ilc and the runtime pack are in one package today. We can solve that by allowing `-buildtoolsruntime` to produce partial package or by unconditionally using `PublishAotUsingRuntimePack` for the live tools runtime.

> The second invocation will first need to undo the first invocation's work like evicting the cache

This can be handled by `-buildtoolsruntime` overriding the output directory in places with conflicts.

am11 commented:
> * Bootstrap community supported platform:
> 
> 
> ```
> ./build.sh -os freebsd -c release -buildtoolsruntime
> ./build.sh -os freebsd -c release
> ```

There are three variants:

1. `dotnet ilc.dll` using live aot runtime build + libs
2. `ilc` published with PublishSingleFile using live or LKG aot runtime + libs
3. `ilc` published with PublishAot using live or LKG aot runtime + libs 

(actually there is another; if we incorporate PublishReadyToRun, but it's tied with PublishingSingleFile)

I think for ILCompiler.csproj (NativeAotSupported=true), we should skip variant 2. AFAIK, variant 2 (which this PR is about) brings unnecessary overhead for ILCompiler; because when we are building ILCompiler, we know that target platform has NativeAotSupported=true, which is different than other stuff like crossgen2_publish.csproj src/native/managed etc. with optional PublishAot or PublishSingleFile. IOW, variant-2 is only meaningful for projects other than ILCompiler in bootstrapping approach.

Then the first command from your example `./build.sh -os freebsd -c release -buildtoolsruntime` will only be dealing with building `ilc.dll` and second command `./build.sh -os freebsd -c release` will be publishing ILCompiler with variant 3 (PublishAot=true); using the cross build mechanism implemented in BuildIntegration.

jkotas commented:
Yes, if the target platform has `NativeAotSupported=true`, `-buildtoolsruntime` needs to build native aot toolset only since that's the only part needed from LKG in that case in a regular build. `-buildtoolsruntime` should build both ilc and matching native aot runtime pack. Otherwise, it would not be possible to mix and match build flavors (my second example).

If the binaries built by the first and second command overlap exactly (it should be the case in the first example), the regular incremental build mechanism should be able to detect that and skip rebuilding them in the second command.

am11 commented:
> Yes, if the target platform has `NativeAotSupported=true`, `-buildtoolsruntime` needs to build native aot toolset only since that's the only part needed from LKG in that case in a regular build.

One problem is that `Build` target on ILCompiler.csproj is triggering `Publish` target, and publishing aot toolset requires freebsd apphost which needs to be published first. AFAICT, without decoupling them (build only and publish only), the first step would need to build the aot runtime and libraries for the second step to complete the package.

`crossgen2.csproj` (for build only), `crossgen2_inbuild.csproj` (for cross build) and `crossgen2_publish.csproj` (for publishing) solve this problem by separating these concerns cleanly: https://github.com/dotnet/runtime/pull/92677#issuecomment-1740066874. ILCompiler.csproj is monolithic in nature compared to crossgen2.

jkotas commented:
> the first step would need to build the aot runtime and libraries for the second step to complete the package.

I think it should work like that even if ILCompiler project is decoupled. Otherwise, my second example (use live checked runtime for release build of the compilers) would not work.

Thefrank commented:
I was using this and the above instructions to produce a FreeBSD ILCompiler package under Linux. This patch no longer seems to work for me when it is applied. 
```
/runtime/.dotnet/sdk/9.0.100-rc.1.24452.12/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(158,5): error NETSDK1084: There is no application host available for the specified RuntimeIdentifier 'freebsd-x64'. [/runtime/Build.proj]
##[error]/runtime/.dotnet/sdk/9.0.100-rc.1.24452.12/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(158,5): error NETSDK1084: (NETCORE_ENGINEERING_TELEMETRY=Restore) There is no application host available for the specified RuntimeIdentifier 'freebsd-x64'.
```

am11 commented:
@Thefrank, while this PR branch doesn't have an issue, merging upstream's `main` to it is failing. It regressed in c4b3d81df13b843ae7bbd8c4b156bd12ccb908f7 as the previous commit, 3227d4ada44f608b47f2e38a39f0fc9646102045, is fine.

am11 commented:
Due to https://github.com/dotnet/msbuild/issues/6391, I can only tell that `packs` subset is causing the restore issue. Perhaps we need for https://github.com/dotnet/sdk/pull/43070 to merge and then flow in runtime's global.json (i.e. couple of weeks at least)?

jkotas commented:
@am11 Would you be interested in implementing the live AOT runtime build option along the lines described above?

am11 commented:
@jkotas, sure. Currently, we have a build break due to net9.0->net10.0 transition https://github.com/dotnet/runtime/pull/105004#issuecomment-2392815193 (even before the LKG PR was checked in). Regardless, I think implementing your proposal will require us to go though those 'how skip apphost package restoration' type of details anyway, so we can start working on it.

Thefrank commented:
nudge

am11 commented:
I've made some progress and pushed to outputrid2 branch https://github.com/dotnet/runtime/compare/main...am11:runtime:feature/freebsd-port/outputrid2, currently it builds clr and fails late in libraries: https://github.com/am11/CrossRepoCITesting/actions/runs/12101359561/job/33741104482. I had to condition couple of places requiring apphost. One of those places is around GenFacades. I think the failures are due to that. Does anyone have a clue how to proceed?

am11 commented:
@ViktorHofer, @akoeplinger, @agocke, @jkoritzinsky, any advice how to proceed here☝️ We are trying to avoid apphost (using `-p:StageOneBuild=true`). Community platforms are blocked on nativeaot and/or crossgen2 builds.

* https://github.com/dotnet/runtime/issues/104497
* https://github.com/dotnet/runtime/issues/108973
* https://github.com/dotnet/runtime/pull/110688
* https://github.com/dotnet/runtime/pull/110687

and possibly on LA64 as well https://github.com/dotnet/runtime/issues/108973#issuecomment-2540813693.

am11 committed:
Add multi-stages for community supported platforms

am11 committed:
Revert to old main

am11 committed:
Merge dotnet/main into feature/freebsd-port/outputrid

am11 commented:
Figured out a slightly different approach while keeping the multi-stage semtantics:

1. build clr+libs with `StageOneBuild=true`, it will skip publishing crossgen2 and ILCompiler.
2. build clr+libs+packs with `StageTwoBuild=true`, it will publish crossgen2 and ILCompiler using the apphost built by the first stage.

@Thefrank commands: https://github.com/am11/CrossRepoCITesting/actions/runs/12387360762/workflow#L24, logs: https://github.com/am11/CrossRepoCITesting/actions/runs/12387360762, artifacts: https://github.com/am11/CrossRepoCITesting/releases/tag/freebsd_12387360762. All ELF binaries in those packages are FreeBSD.

@dotnet/samsung, also tested linux-rsicv64 Crossgen2 and NativeAOT (applying https://github.com/dotnet/runtime/pull/110688) builds using the same commands:
 `build.sh clr+libs -coss -a riscv64 -p:StageOneBuild=true`
 `build.sh clr+libs+host -coss -a riscv64 -p:StageTwoBuild=true`

 If this gets merged before #110688, I will push a one liner patch enabling riscv64 in `_NativeAotSupportedArch`. Then we can test and perfect nativeaot on linux-riscv64 in isolation.

> @am11 Would you be interested in implementing the live AOT runtime build option along the lines described above?

@jkotas PTAL, hopefully this is by the outlined specs.

jkotas commented on src/coreclr/tools/aot/crossgen2/crossgen2.props:
```diff
@@ -15,9 +15,11 @@
     <EnableDefaultEmbeddedResourceItems>false</EnableDefaultEmbeddedResourceItems>
     <Configurations>Debug;Release;Checked</Configurations>
     <RunAnalyzers>false</RunAnalyzers>
+    <UseLocalTargetingRuntimePack Condition="'$(StageTwoBuild)' == 'true'">true</UseLocalTargetingRuntimePack>
```
Can this be in `toolAot.targets` so that it applies to all tools published with Aot automatically?

am11 replied:
We can try :)

```suggestion
```

am11 replied:
Unfortunately, it is giving errors like `/__w/1/s/src/libraries/Directory.Build.targets(120,3): error MSB4011: "/__w/1/s/eng/targetingpacks.targets" cannot be imported again. It was already imported at "/__w/1/s/eng/toolAot.targets (7,3)". This is most likely a build authoring error. This subsequent import will be ignored. [/__w/1/s/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Microsoft.Interop.SourceGeneration.csproj] [/__w/1/s/Build.proj]`

We will need to add tracking. targetingpacks.targets is used in 29 places. I'm leaving it as is for now.

ViktorHofer replied:
A future change should import targetingpacks.targets in the repo root un-conditionally and enable the features in it via feature knobs. I should have done that in the first place back when I introduced the file.

jkotas commented on eng/Subsets.props:
```diff
@@ -124,7 +124,8 @@
     <_NativeAotSupportedArch Condition="'$(TargetArchitecture)' == 'x64' or '$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm' or '$(TargetArchitecture)' == 'loongarch64' or ('$(TargetOS)' == 'windows' and '$(TargetArchitecture)' == 'x86')">true</_NativeAotSupportedArch>
     <NativeAotSupported Condition="'$(_NativeAotSupportedOS)' == 'true' and '$(_NativeAotSupportedArch)' == 'true'">true</NativeAotSupported>
-    <UseNativeAotForComponents Condition="'$(NativeAotSupported)' == 'true' and '$(TargetOS)' == '$(HostOS)' and '$(TargetsLinuxBionic)' != 'true'">true</UseNativeAotForComponents>
+    <_IsCommunityArchitecture Condition="'$(CrossBuild)' == 'true' and ('$(TargetArchitecture)' == 'loongarch64' or '$(TargetArchitecture)' == 'riscv64')">true</_IsCommunityArchitecture>
```
I would expect the community architectures to be able to `UseNativeAotForComponents` here if they have otherwise working NAOT support.

What was wrong with the original condition?

am11 replied:
This is in line with what we have been using thus far.

1. linux-loongarch64 folks are not using -cross so they have not observed this issue https://github.com/dotnet/runtime/issues/108973#issuecomment-2541077363
2. freebsd-x64 is protected by TargetOS == HostOS condition below

When building on the device (CrossBuild=false) with previous SDK version from global.json, there is no problem and we don't need multi-stages. Otherwise, we will need to add stage3 as stage2 is the first one publishes ILCompiler.csproj.

jkotas replied:
We only need an ILCompiler that runs on host. What prevents it from being produced in the first stage?

am11 replied:
Stage 1 does not publish ILCompiler and crossgen2 because they need live apphost build. Stage 2 publishes them as single file using live apphost from stage 1.

am11 commented on src/coreclr/tools/aot/crossgen2/crossgen2.props:
```diff
@@ -15,9 +15,11 @@
   </PropertyGroup>
 
   <Import Project="../AotCompilerCommon.props" />
+  <Import Project="$(RepositoryEngineeringDir)targetingpacks.targets" Condition="'$(StageTwoBuild)' == 'true'" />
```
```suggestion
```

am11 commented on src/coreclr/tools/aot/ILCompiler/ILCompiler.props:
```diff
@@ -12,9 +12,11 @@
     <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
     <EnableDefaultEmbeddedResourceItems>false</EnableDefaultEmbeddedResourceItems>
     <Configurations>Debug;Release;Checked</Configurations>
+    <UseLocalTargetingRuntimePack Condition="'$(StageTwoBuild)' == 'true'">true</UseLocalTargetingRuntimePack>
```
```suggestion
```

am11 commented on src/coreclr/tools/aot/ILCompiler/ILCompiler.props:
```diff
@@ -12,9 +12,11 @@
   </PropertyGroup>
 
   <Import Project="../AotCompilerCommon.props" />
+  <Import Project="$(RepositoryEngineeringDir)targetingpacks.targets" Condition="'$(StageTwoBuild)' == 'true'" />
```
```suggestion
```

am11 committed:
Apply suggestions from code review

am11 committed:
Update toolAot.targets

am11 committed:
Revert experiment

am11 committed:
Merge branch 'main' into feature/freebsd-port/outputrid

am11 commented on eng/pipelines/runtime.yml:
```diff
@@ -493,8 +493,11 @@ extends:
-            buildArgs: -s clr+libs+host -c $(_BuildConfig) -rc Checked
+            buildArgs: -s clr+libs+host -c $(_BuildConfig) -rc Checked -p:StageOneBuild=true
             timeoutInMinutes: 120
+            postBuildSteps:
+              - script: $(Build.SourcesDirectory)/build$(scriptExt) ${{ variables.debugOnPrReleaseOnRolling }} -s clr.tools+packs -c $(_BuildConfig) -rc Checked -cross -os freebsd -p:StageTwoBuild=true
+                displayName: Build clr.tools and packs
```
```suggestion
            postBuildSteps:
              - script: |
                  platform=${{ parameters.platforms }}
                  IFS='_' read -r os arch <<< "$platform"

                  $(Build.SourcesDirectory)/build$(scriptExt) \
                    -os $os                                   \
                    -arch $arch                               \
                    -c $(_BuildConfig)                        \
                    -s clr.tools+packs -rc Checked -cross
                displayName: Build clr.tools and packs
```

am11 committed:
Replace hardcoded platform arg

am11 commented:
@jkotas, @jkoritzinsky, community platform builds are green. Note that `freebsd-x64` and `linux-loongarch64` set `NativeAotSupported` as before. If https://github.com/dotnet/runtime/pull/110688 is merged before this PR, then we can add `targetArchitecture==riscv64` condition in Subsets.props and enable it. Otherwise, I will add it in that PR or as a separate one once both are merged.

If we want to enable `UseNativeAotForComponents` in the CI, then we will need to add a third stage; first stage prepares apphost, second stage uses `PublishSingleFile` for tools build and the third stage will use tools from second stage to use `PublishAot`. Please let me know if we want the third stage.

jkotas commented on src/coreclr/tools/aot/ILCompiler/ILCompiler.props:
```diff
@@ -12,9 +12,11 @@
     <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
     <EnableDefaultEmbeddedResourceItems>false</EnableDefaultEmbeddedResourceItems>
     <Configurations>Debug;Release;Checked</Configurations>
+    <UseLocalTargetingRuntimePack Condition="'$(StageTwoBuild)' == 'true'">true</UseLocalTargetingRuntimePack>
```
Is this a problem for incremental builds? Should the build with LKG target and the build local targeting pack go into separate directories?

am11 replied:
This project is not built in stage 1, so incremental build does not come into play.

jkotas commented:
I am not following why we would need 3 stages to enable UseNativeAotForComponents. The tools built by the first stage need to target the build machine architecture only to unblock second stage. We should have everything available to build the tools that target the build machine architecture during the first stage.

am11 commented on eng/Subsets.props:
```diff
@@ -121,11 +121,11 @@
-    <NativeAotSupported Condition="'$(_NativeAotSupportedOS)' == 'true' and '$(_NativeAotSupportedArch)' == 'true' and '$(_IsCommunityCrossArchitecture)' != 'true'">true</NativeAotSupported>
-    <UseNativeAotForComponents Condition="'$(NativeAotSupported)' == 'true' and '$(TargetOS)' == '$(HostOS)' and '$(TargetsLinuxBionic)' != 'true'">true</UseNativeAotForComponents>
+    <NativeAotSupported Condition="'$(_NativeAotSupportedOS)' == 'true' and '$(_NativeAotSupportedArch)' == 'true'">true</NativeAotSupported>
+    <_IsCommunityCrossArchitecture Condition="'$(CrossBuild)' == 'true' and ('$(TargetArchitecture)' == 'loongarch64' or '$(TargetArchitecture)' == 'riscv64')">true</_IsCommunityCrossArchitecture>
+    <UseNativeAotForComponents Condition="'$(NativeAotSupported)' == 'true' and '$(TargetOS)' == '$(HostOS)' and '$(TargetsLinuxBionic)' != 'true' and '$(_IsCommunityCrossArchitecture)' != 'true'">true</UseNativeAotForComponents>
```
```suggestion
    <UseNativeAotForComponents Condition="'$(NativeAotSupported)' == 'true' and ('$(TargetOS)' == '$(HostOS)' or '$(HostOS)' != 'windows') and '$(TargetsLinuxBionic)' != 'true'">true</UseNativeAotForComponents>
```

am11 committed:
Try with UseNativeAotForComponents=true

ViktorHofer commented on eng/targetingpacks.targets:
```diff
@@ -164,7 +164,9 @@
     <ItemGroup>
       <RuntimeFramework Version="$(ProductVersion)"
-                        Condition="'%(RuntimeFramework.FrameworkName)' == '$(LocalFrameworkOverrideName)'" />
+                        Condition="'$(StageTwoBuild)' != 'true' and '%(RuntimeFramework.FrameworkName)' == '$(LocalFrameworkOverrideName)'" />
+      <RuntimeFramework Version="$(BundledNETCoreAppPackageVersion)"
+                        Condition="'$(StageTwoBuild)' == 'true' and '%(RuntimeFramework.FrameworkName)' == '$(LocalFrameworkOverrideName)'" />
```
Apply the condition directly on the Version attribute so that you don't need two items:

```
<RuntimeFramework Condition="'%(RuntimeFramework.FrameworkName)' == '$(LocalFrameworkOverrideName)'">
  <Version Condition="'$(StageTwoBuild)' != 'true'">$(ProductVersion)</Version>
  <Version Condition="'$(StageTwoBuild)' == 'true'">$(BundledNETCoreAppPackageVersion)</Version>
</RuntimeFramework>
```

am11 committed:
Apply condition on versino directly

am11 commented on eng/Subsets.props:
```diff
@@ -121,11 +121,10 @@
-    <NativeAotSupported Condition="'$(_NativeAotSupportedOS)' == 'true' and '$(_NativeAotSupportedArch)' == 'true' and '$(_IsCommunityCrossArchitecture)' != 'true'">true</NativeAotSupported>
-    <UseNativeAotForComponents Condition="'$(NativeAotSupported)' == 'true' and '$(TargetOS)' == '$(HostOS)' and '$(TargetsLinuxBionic)' != 'true'">true</UseNativeAotForComponents>
+    <NativeAotSupported Condition="'$(_NativeAotSupportedOS)' == 'true' and '$(_NativeAotSupportedArch)' == 'true'">true</NativeAotSupported>
+    <UseNativeAotForComponents Condition="'$(NativeAotSupported)' == 'true' and ('$(TargetOS)' == '$(HostOS)' or '$(HostOS)' != 'windows') and '$(TargetsLinuxBionic)' != 'true'">true</UseNativeAotForComponents>
```
```suggestion
    <UseNativeAotForComponents Condition="'$(NativeAotSupported)' == 'true' and '$(TargetsLinuxBionic)' != 'true'">true</UseNativeAotForComponents>
```

am11 committed:
Remove over-defensive condition

This was only needed for Android target on Windows build machine, which is guarded by TargetsLinuxBionic condition.

am11 committed:
Revert UseNativeAotForComponents change

am11 commented:
> I am not following why we would need 3 stages to enable UseNativeAotForComponents. The tools built by the first stage need to target the build machine architecture only to unblock second stage. We should have everything available to build the tools that target the build machine architecture during the first stage.

@jkotas, I have tested it. Setting `UseNativeAotForComponents` alone did not use live AOT build as expected, it just published crossgen2/ilc with corehost. Seems like it requires more configurations like we have in `tests.singlefile.targets` etc. so AOT toolchain can find `PrivateSdkAssemblies` and `FrameworkAssemblies`.

am11 commented on eng/Subsets.props:
```diff
@@ -360,9 +360,9 @@
                              $(CoreClrProjectRoot)tools\AssemblyChecker\AssemblyChecker.csproj" Category="clr" Condition="'$(DotNetBuildSourceOnly)' != 'true'"/>
     <!-- skip the architectures that don't have LKG runtime packs -->
-    <ProjectToBuild Include="$(CoreClrProjectRoot)tools\aot\crossgen2\crossgen2_publish.csproj" Condition="'$(NativeAotSupported)' == 'true'" Category="clr" />
+    <ProjectToBuild Include="$(CoreClrProjectRoot)tools\aot\crossgen2\crossgen2_publish.csproj" Condition="'$(NativeAotSupported)' == 'true' and '$(StageOneBuild)' != 'true'" Category="clr" />
```
@jkoritzinsky `ILCompiler_publish.csproj` is not listed in this file (or anywhere else). Was it an oversight or something for the future?

jkoritzinsky replied:
I think there was an intent to hook it up for the `Microsoft.DotNet.ILCompiler` package, but I think it didn't quite happen. Looks like we still publish from the `ILCompiler.csproj` project and ship that. I think the build needs to be adjusted to hook this up correctly (use `ILCompiler_publish.csproj` to publish ilc and remove the publishing logic from `ILCompiler.csproj`)

am11 commented:
> `PublishAotUsingRuntimePack`

It gives

```
/runtime/.dotnet/sdk/10.0.100-alpha.1.24610.7/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(497,5):
error NETSDK1082: There was no runtime pack for Microsoft.NETCore.App available for the specified RuntimeIdentifier 'linux-loongarch64'.
[/runtime/src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj]
```

We can look into `UseNativeAotForComponents` separately. Original issue https://github.com/dotnet/runtime/issues/104497 was about corehost published `ilc` binary is using linux apphost on freebsd cross build, which is fixed.

am11 committed:
Merge branch 'main' into feature/freebsd-port/outputrid

am11 commented on eng/Subsets.props:
```diff
@@ -121,11 +121,11 @@
     <!-- CLR NativeAot only builds in a subset of the matrix -->
-    <_IsCommunityCrossArchitecture Condition="'$(CrossBuild)' == 'true' and ('$(TargetArchitecture)' == 'loongarch64' or '$(TargetArchitecture)' == 'riscv64')">true</_IsCommunityCrossArchitecture>
     <_NativeAotSupportedOS Condition="'$(TargetOS)' == 'windows' or '$(TargetOS)' == 'linux' or '$(TargetOS)' == 'osx' or '$(TargetOS)' == 'maccatalyst' or '$(TargetOS)' == 'iossimulator' or '$(TargetOS)' == 'ios' or '$(TargetOS)' == 'tvossimulator' or '$(TargetOS)' == 'tvos' or '$(TargetOS)' == 'freebsd'">true</_NativeAotSupportedOS>
     <_NativeAotSupportedArch Condition="'$(TargetArchitecture)' == 'x64' or '$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm' or '$(TargetArchitecture)' == 'loongarch64' or ('$(TargetOS)' == 'windows' and '$(TargetArchitecture)' == 'x86')">true</_NativeAotSupportedArch>
```
```suggestion
    <_NativeAotSupportedArch Condition="'$(TargetArchitecture)' == 'x64' or '$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm' or '$(TargetArchitecture)' == 'loongarch64' or '$(TargetArchitecture)' == 'riscv64' or ('$(TargetOS)' == 'windows' and '$(TargetArchitecture)' == 'x86')">true</_NativeAotSupportedArch>
```

am11 committed:
Enable NativeAot runtime build on riscv64

am11 commented on eng/Subsets.props:
```diff
@@ -121,11 +121,11 @@
-    <_NativeAotSupportedArch Condition="'$(TargetArchitecture)' == 'x64' or '$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm' or '$(TargetArchitecture)' == 'loongarch64' or ('$(TargetOS)' == 'windows' and '$(TargetArchitecture)' == 'x86')">true</_NativeAotSupportedArch>
-    <NativeAotSupported Condition="'$(_NativeAotSupportedOS)' == 'true' and '$(_NativeAotSupportedArch)' == 'true' and '$(_IsCommunityCrossArchitecture)' != 'true'">true</NativeAotSupported>
-    <UseNativeAotForComponents Condition="'$(NativeAotSupported)' == 'true' and '$(TargetOS)' == '$(HostOS)' and '$(TargetsLinuxBionic)' != 'true'">true</UseNativeAotForComponents>
+    <_NativeAotSupportedArch Condition="'$(TargetArchitecture)' == 'x64' or '$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm' or '$(TargetArchitecture)' == 'loongarch64' or '$(TargetArchitecture)' == 'riscv64' or ('$(TargetOS)' == 'windows' and '$(TargetArchitecture)' == 'x86')">true</_NativeAotSupportedArch>
```
I think we should invert this condition; blacklist instead of whitelist the unsupported platforms.

am11 commented:
This is also needed to test riscv64 AOT https://github.com/dotnet/runtime/issues/106223#issuecomment-2582967863.

am11 committed:
Merge branch 'main' into feature/freebsd-port/outputrid

am11 committed:
Update Subsets.props

am11 commented:
@jkotas, if this looks ok, can this be merged?

akoeplinger commented:
@am11 the build failures on freebsd/loongarch/riscv look related

am11 committed:
merge conflicts..

am11 committed:
Merge branch 'main' into feature/freebsd-port/outputrid

jkotas commented:
> @jkotas, if this looks ok, can this be merged?

I was hoping to get proper two stage (cross-plat) build for native AOT.

If the UseLocalTargetingRuntimePack changes look fine to @ViktorHofer, this can be merged. I expect that we will want to do more work on this over time.

ViktorHofer commented on eng/pipelines/runtime.yml:
```diff
@@ -493,8 +493,26 @@ extends:
             timeoutInMinutes: 120
+            postBuildSteps:
+              - script: |
+                  echo Running $(Build.SourcesDirectory)/build$(scriptExt)  \
+                    ${{ variables.debugOnPrReleaseOnRolling }}              \
+                    -s clr.tools+packs -rc Checked -cross                   \
+                    -os $(osGroup)                                          \
+                    -a $(archType)                                          \
+                    -c $(_BuildConfig)                                      \
+                    -p:StageTwoBuild=true
```
This echo feels a bit weird to have here.

am11 replied:
It was added to show the command. Apparently logs don't capture postBuildSteps/script command without it.

ViktorHofer commented on eng/targetingpacks.targets:
```diff
@@ -163,8 +163,10 @@
-      <RuntimeFramework Version="$(ProductVersion)"
-                        Condition="'%(RuntimeFramework.FrameworkName)' == '$(LocalFrameworkOverrideName)'" />
+      <RuntimeFramework Condition="'%(RuntimeFramework.FrameworkName)' == '$(LocalFrameworkOverrideName)'">
+        <Version Condition="'$(StageTwoBuild)' != 'true'">$(ProductVersion)</Version>
+        <Version Condition="'$(StageTwoBuild)' == 'true'">$(BundledNETCoreAppPackageVersion)</Version>
```
Just to better understand this change, we aren't going to use this new switch anywhere in our own product composition? This is just for some community flavours? We don't have any support for this in the VMR and I don't think would want this initially.

am11 replied:
Yup, this is only when user explicitly passes `-p:Stage{One,Two}Build=true` on command-line for community platforms, where target apphost package is not available during the crossbuild.

An alternative way is that we ask the maintainer to provide the packages for the target version of apphost, crossgen2 and ilcompiler. e.g. @Thefrank has GitHub actions set up to build dotnet/sdk for FreeBSD which publishes the SDK to GitHub releases and nuget pacakges to GitHub's nuget feed. It uses `-p:OfficialBuildId=<timestamp version>` same as Microsoft official builds. We can make those workflows trigger based, i.e. when the daily SDK is published, trigger the workflow for the same version (or just run workflow on-demand when we need to bump the SDK in global.json). This way, community platforms will also be on the LKG plan: https://github.com/dotnet/runtime/pull/107772#discussion_r1769508628. We can add their feeds in NuGet.config in runtime repo. Or maybe even a dedicated repository `github.com/dotnet/commuty-paltforms` with just the GitHub workflow and scripts for publishing would be appropriate. Prior art is all figured out: https://github.com/Thefrank/dotnet-freebsd-crossbuild, https://github.com/am11/CrossRepoCITesting/tree/107b1152fc6acd892bafc23c4eb84f0387f9f4ca/.github/workflows

am11 replied:
Yup, this is only when user explicitly passes `-p:Stage{One,Two}Build=true` on command-line for community platforms, where target LKG apphost package is not available during the crossbuild.

An alternative way is that we ask the maintainer to provide the packages for the target version of apphost, crossgen2 and ilcompiler. e.g. @Thefrank has GitHub actions set up to build dotnet/sdk for FreeBSD which publishes the SDK to GitHub releases and nuget pacakges to GitHub's nuget feed. It uses `-p:OfficialBuildId=<timestamp version>` same as Microsoft official builds. We can make those workflows trigger based, i.e. when the daily SDK is published, trigger the workflow for the same version (or just run workflow on-demand when we need to bump the SDK in global.json). This way, community platforms will also be on the LKG plan: https://github.com/dotnet/runtime/pull/107772#discussion_r1769508628. We can add their feeds in NuGet.config in runtime repo. Or maybe even a dedicated repository `github.com/dotnet/commuty-paltforms` with just the GitHub workflow and scripts for publishing would be appropriate. Prior art is all figured out: https://github.com/Thefrank/dotnet-freebsd-crossbuild, https://github.com/am11/CrossRepoCITesting/tree/107b1152fc6acd892bafc23c4eb84f0387f9f4ca/.github/workflows. This way, we won't need the multi-stage build.

jkotas replied:
> An alternative way is that we ask the maintainer to provide the packages for the target version of apphost, crossgen2 and ilcompiler

My preference would be that that StageOne build produces these packages, puts them into a directory under artifacts and then StageTwo build fetches these packages from that directory. It is how Roslyn two stage build works: https://github.com/dotnet/roslyn/blob/main/docs/contributing/bootstrap-builds.md

ViktorHofer replied:
From that document:

> The build correctness leg ensures that the latest code in dotnet/roslyn can be used to build dotnet/roslyn. Essentially our compiler can bootstrap itself.

Sounds like they are doing this for validation and not produce shipping artifacts. If the initial incentive here is validation as well then we should perform this at the VMR level. We already have infrastructure for that (source-build has a stage1 and stage2 build). At the VMR level we would be able to validate the entire .NET SDK, not just one or a few repositories.

am11 replied:
> My preference would be that that StageOne build produces these packages

StageOne doesn't have apphost for target, that's why package is produced in StageTwo. The commit in this PR was changing the order of build (which was rejected); that was building apphost before the package.

am11 replied:
> My preference would be that that StageOne build produces these packages

StageOne doesn't have apphost for target, that's why package is produced in StageTwo. The first commit in this PR was changing the order of build (which was rejected); that was building apphost before the package step.

jkotas replied:
> We already have infrastructure for that (source-build has a stage1 and stage2 build). 

The source build stage1 build expects LKG .NET runtime for the target platform today. Do you expect that we are going to fix source build to not depend on LKG .NET runtime for the target platform? I believe that it would require the two-stage runtime build that this PR is about.

It is a generic capability that can be used for number of purposes:
- Community platforms may choose to use it to build shipping bits
- In addition to basic validation that I agree can be done in VMR, it would be useful to have a validation for more advanced configurations like checked compilers used to compile the compilers + release compilers (ie checked Stage 1 build and release State 2 build). This configuration is useful to find bugs. I have done it number of times manually. I would love to have this capability built in so that the configurations like this can be run in dotnet/runtime CI (e.g. as outer loop jobs). I do not think it would be appropriate for VMR.

jkotas replied:
> StageOne doesn't have apphost for target, that's why package is produced in StageTwo. 

I am sorry but I still do not see why it is necessary. StageOne needs to build compilers that run on the build machine. It should not require apphost for target.

am11 replied:
(title+description of this PR) There are two issues:

1. When target OS mismatch and target arch is the same, then the issue is silent from top post https://github.com/dotnet/runtime/pull/105004#issue-2412339511 we get linux ELF on freebsd.

2. When target arch mismatches,, then it tries to downloads for `linux-{arch}` (linux- is hardcoded which is wrong..), and since the nuget feed doesn't ahve riscv64 and loongarch64 archs, we get restore time errors.

am11 replied:
(title+description of this PR) There are two issues:

1. When target OS mismatches and target arch is the same, then the issue is silent from top post https://github.com/dotnet/runtime/pull/105004#issue-2412339511 we get linux ELF on freebsd.

2. When target arch mismatches, then it tries to downloads for `linux-{arch}` (linux- is hardcoded which is wrong..), and since the nuget feed doesn't have riscv64 and loongarch64 archs, we get restore time errors.

ViktorHofer replied:
> The source build stage1 build expects LKG .NET runtime for the target platform today. Do you expect that we are going to fix source build to not depend on LKG .NET runtime for the target platform?

I didn't mean that we would run this in a source-only configuration. I meant that thanks for source-build we already have infrastructure in the VMR. We should be able to "simply" enable it for the msft builds there. The msft builds use an LKG SDK that targets the host and not the target (i.e. when building the win-x86 vertical we use a win-x64 host SDK to produce the win-x86 targets). Just wanted to clarify this.

If this PR provides more than what the VMR could validate, sure, go for it. My primary concern was around adding new infrastructure that we don't exercise anywhere in the VMR. But given that this is for runtime repo only validation, you can dismiss my concerns.

ViktorHofer reviewed (commented):
@jkoritzinsky please also take a look

am11 commented:
> I was hoping to get proper two stage (cross-plat) build for native AOT.

@jkotas, your ask was a two stage solution with NativeAotSupported+UseNativeAotForComponents, I have provided two stage solution with NativeAotSupported and sufficiently explained why UseNativeAotForComponents part is not an option by sharing what I tried and how it failed (also visible in the CI). Either it is possible and I have no idea how to go about it, or it is not possible and I am having a hard time explaining.

In any case, this is getting a bit out of scope for this PR as it was supposed to be fixing https://github.com/dotnet/runtime/issues/104497.

> I expect that we will want to do more work on this over time.

Definitely, like everything this can improve as well. Now that CI is capturing the usage pattern of Samsung and FreeBSD folks, it would be slightly harder to regress it and cause regressions in the first place.

I suggest we merge this PR to avoid continuing merge conflicts.

jkoritzinsky commented on src/coreclr/tools/aot/crossgen2/crossgen2_inbuild.csproj:
```diff
@@ -12,9 +12,6 @@
     -->
     <PublishAot Condition="$(NETCoreSdkPortableRuntimeIdentifier.StartsWith('win-'))">true</PublishAot>
     <PublishSingleFile Condition="'$(PublishAot)' != 'true'">true</PublishSingleFile>
-    <PublishReadyToRun>true</PublishReadyToRun>
-    <PublishReadyToRunComposite>true</PublishReadyToRunComposite>
-    <TrimmerSingleWarn>false</TrimmerSingleWarn>
```
Can we change these to be conditioned on `'$(StageOneBuild)' == ''`? This will cause a pretty large build time regression in the VMR now that we use this across the stack to build the product (and not just in dotnet/runtime).

am11 replied:
Can we use PublishAot=true as before? I think this was using PublishAot=true via `eng/toolAot.targets` before https://github.com/dotnet/runtime/pull/110676.

jkoritzinsky replied:
We can't as we may not have the libraries necessary to AOT available (or a toolchain with enough pieces available) to AOT for the host machine in our Azure Linux 3.0 cross build docker images.

am11 replied:
I am using AZ 3.0 prereq images, all dependencies seem to be there for host (in `/usr`) and target (in `/crossrootfs/$rid/usr`). `-p:SysRoot=/crossrootfs/$rid` takes care the rest with the host toolchain (dotnet and clang).

jkoritzinsky replied:
When I tried it, there were pieces missing in `/usr` for linking an ILC'd crossgen2 to target the host. We can't use `-p:SysRoot=/crossrootfs/$arch` for the host crossgen2 as it needs to build to target the host machine (ie we can't NAOT crossgen2_inbuild for the x64 host machine using an arm64 crossrootfs).

am11 replied:
For ilc, I'm using this method https://github.com/dotnet/runtime/issues/106223#issuecomment-2582967863. After the stages build, I built ilc for host then used it to publish for target. It downloads the runtime pack for target and uses host ilc. Once we have SDK updated which recognizes linux-riscv64 as a valid platform for ILCompiler bundle, it will exempt this step (among other things):

> 4. copy `libclrjit_unix_riscv64_arm64.so` from `/runtime/artifacts/bin/coreclr/linux.riscv64.Checked/arm64/libclrjit_unix_riscv64_arm64.so` to `/runtime/artifacts/bin/coreclr/linux.arm64.Debug/ilc-published/`

jkoritzinsky replied:
To avoid making the multistage build a requirement for all platforms, can we please change this to use R2R in the regular build flow instead of deleting it?

am11 replied:
The multistage is only required on community platform because we don’t have runtime packs in nuget feed. Officially supported platforms download them as part of the restore.

am11 replied:
The multistage is only required on community platform because we don’t have target runtime packs in nuget feed. Officially supported platforms download them as part of the restore.

jkoritzinsky replied:
Okay. In any case, when I tried to publish with AOT for the host machine on the Azl3 images in the VMR, it failed with missing libraries. Can we please set this up to use R2R when supported (ie not community platforms)?

Once we have this fixed, we can merge this in.

am11 replied:
>  it failed with missing libraries

Do you remember which missing library was it when you were using the shared target specifically with this line https://github.com/dotnet/runtime/blob/6e1486e7bf62d4268c198833fa3590b8f79cd99d/eng/toolAot.targets#L53 target? https://github.com/dotnet/runtime/pull/110676 was merged only five days ago and we have been using this stuff for a while. I have just tested it on `mcr.microsoft.com/dotnet-buildtools/prereqs:azurelinux-3.0-net10.0-cross-arm64` and it publishes AOT app for x64 (host) just fine.

am11 replied:
>  it failed with missing libraries

Do you remember which missing library was it when you were using the shared target specifically with this line https://github.com/dotnet/runtime/blob/6e1486e7bf62d4268c198833fa3590b8f79cd99d/eng/toolAot.targets#L53 https://github.com/dotnet/runtime/pull/110676 was merged only five days ago and we have been using this stuff for a while. I have just tested it on `mcr.microsoft.com/dotnet-buildtools/prereqs:azurelinux-3.0-net10.0-cross-arm64` and it publishes AOT app for x64 (host) just fine; of course with LinkerFlavor=lld.

jkoritzinsky replied:
I don't remember off the top of my head. It was a few weeks ago when I started that PR that I hit this issue. We can try to do a test PR into the VMR to make sure it would work when it flows. If it would, then we can change it here.

am11 replied:
Yup, it would be nice to use PublishAot'd crossgen2 and ilc on non-community paltforms because that's the version of binary what the end-use gets via nuget package. For now, I will make it conditional on !multistage. 🙂

jkoritzinsky commented on eng/pipelines/runtime.yml:
```diff
@@ -493,8 +493,26 @@ extends:
           jobParameters:
             testScope: innerloop
             nameSuffix: CoreCLR
```
```suggestion
            nameSuffix: CoreCLR_TwoStage
```

am11 committed:
Update eng/pipelines/runtime.yml

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

am11 committed:
Merge branch 'main' into feature/freebsd-port/outputrid

am11 committed:
Skip PublishR2R only on multistage

jkoritzinsky commented on src/coreclr/tools/aot/crossgen2/crossgen2_inbuild.csproj:
```diff
@@ -12,9 +12,14 @@
+    <TrimmerSingleWarn>false</TrimmerSingleWarn>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(CrossBuild)' == 'true' and '$(StageOneBuild)' != 'true' and '$(StageTwoBuild)' != 'true'">
```
```suggestion
  <PropertyGroup Condition="'$(StageOneBuild)' != 'true' and '$(StageTwoBuild)' != 'true'">
```
We can still end up building this for the vertical in what runtime would not consider a cross-build (technically we build it in source-build too).

am11 replied:
Multistage is not designed for source-build or tested with SB at all. Lets fix the issue at hand and worry about enabling multistage for source-build (which highly doubt is necessary) separately.

jkoritzinsky replied:
The problem is the existing condition won't be satisfied when CrossBuild is not true, and it needs to be.

am11 committed:
Update crossgen2_inbuild.csproj

jkoritzinsky commented:
/ba-g wasm failures and timeouts unrelated
